rfcs/0125-bootspec.md

```diff
@@ -0,0 +1,218 @@
+      "<name>": {
+        // A full Bootspec v1 document, _without_ the `"v1"` key.
+        // Note that it is not valid for a specialisation to have further specialisations.
+      }
+    }
+  }
+}
+```
+
+### Risks
+[risks]: #risks
+
+- Some of the bootloader backends are quite complicated, and in many cases have inadequate tests.
+  We could accidentally break corner cases.
+- The bootloader backends are inherently a weak point for NixOS, as it is our last option for rolling back.
+  We cannot roll back a broken bootloader.
+  This and the previous point are risks, but also help demonstrate the value of reducing the amount of code and complexity in the generator.
+
+### Milestones
+[milestones]: #milestones
+
+- Create and package the backwards compatibility synthesizer as a standalone tool.
+  A version of it already exists, but it is not standalone.
+- Generate the bootspec files as part of building the system closure.
+- Update the bootloader backends to use bootspec as their primary source of installation data.
+- Implement a NixOS module which supports external bootloader tooling.
+
+# FAQ
+[faq]: #faq
+
+**Does this involve patching any of the bootloaders?**
+
+No: this work is all about creating a cleaner interface for the tools we maintain which generate the files and configuration for our supported bootloaders.
+
+**Why can't we solve this with the module system?**
+
+Boot-menu generation is by definition cross-generation.
+Bootloader backends, by definition reconfigure the boot partition and records, and are supposed to apply retroactively to old configurations.
+This regeneration happens way beyond the lifecycle of the module system.
+
+**How is this easier than `if (-f "$out/append-initrd-secrets")`?**
+
+The main problem with that approach is there are a lot of files that you need to look for and examine in specific ways.
+We end up duplicating a lot of work across bootloader backends to "learn" about the system.
+By creating a source of truth for this information the bootloader backends can focus more on the work that is unique to their problem.
+
+Doing this work now may also make it easier to refactor a NixOS system's `$out` , reducing the number of symlinks and "noise" if we wanted to do that.
+
+**Does this interrupt the work to make stage-1 based on systemd?**
+
+No.
+Stage-1 won't use this document at all.
+
+**What about memtest and other boot options?**
+
+That should be left to configuration passed to the bootloader backend.
+There could be any number of "extra" bootables like this, and importantly they have nothing to do with a specific NixOS system closure.
+Concretely, a user who enables memtest probably wants the most recent memtest, and not one memtest from every generation.
+
+# Open Questions
+[open-questions]: #open-questions
+
+- Should there be a general-purpose "meta" or "extensions" field which allows arbitrary extension of the data? What should this look like?
+
```

Comment:
```suggestion

- Should there be a non-derivation metadata facility that carries useful metadata for which it isn't desirable to cause rebuilds, redeployments or new boot entries - specifically data such a git commit hash or a date of last modification.

```

Reply:
@roberth Could you clarify what you envision by a "non derivation metadata facility"? Would that be a generalization of bootspec so that bootspec would be only a special case of it? (my interpretation)

We discussed it during our meeting, but I think we had some different interpretations of your comments.

Reply:
By adding a field such as the git revision or last modified date, you're creating new generations that are otherwise identical to preceding ones when you make changes to unrelated files. My open question aims to prevent that from happening.
(The need for this came up in https://github.com/NixOS/nixpkgs/pull/179772)

> Would that be a generalization of bootspec so that bootspec would be only a special case of it?

Possibly.

> > non-derivation

I suppose the metadata _can_ be in a derivation, but this derivation should not be used for determining the equivalence of generations. Assuming we keep using the output path of `system.build.toplevel` for the purpose of generation equivalence, the metadata could be put in a separate derivation that isn't part of `system.build.toplevel`.
There's a number of ways you could implement this sort of thing.

Making `toplevel` not top level isn't great. I could rephrase this as

```suggestion

- How should we identify generations as being equivalent? This will help keep boot menus concise and allows activation to be cut short, especially on systems with frequently changing metadata, such as the revision of a repository that contains components not deployed to the system.
```

This allows other solutions such as a "fingerprint" field containing a hash of the essential values so that duplicate generations can be detected by the bootloader menu generator (or other tool that benefits from a "shortcutting" behavior).

Implementation idea (not necessarily part of bootspec; more of a NixOS implementation detail. `fingerprint` existence and type (string) _is_ bootspec though.)
```
let spec0 = config.system.build.bootspec; # assuming this is where NixOS puts the almost-final bootspec stuff
     specHash = builtins.hashString (builtins.toJSON (config.system.build.bootspec // { meta = null; }));
     spec1 = spec0 // { fingerprint = specHash; };
in {
  # assuming somethign like https://github.com/NixOS/nixpkgs/pull/148456/commits/4ec415cff9bcbaef08e0d900406a5c7181e71881
  config.system.systemBuilderCommands = ''
    ln -s ${json.generate "bootspec.json" spec1} $out/bootspec.json
  '';
}
```